src/main/java/net/mcreator/blockly/java/blocks/ProcedureCallBlock.java

```diff
@@ -30,60 +30,78 @@
 import net.mcreator.util.XMLUtil;
 import org.w3c.dom.Element;
 
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
+import java.util.stream.Collectors;
 
 public class ProcedureCallBlock implements IBlockGenerator {
 
 	@Override public void generateBlock(BlocklyToCode master, Element block) throws TemplateGeneratorException {
-		Element element = XMLUtil.getFirstChildrenWithName(block, "field");
+		Element procedureField = XMLUtil.getFirstChildrenWithName(block, "field");
 		String type = block.getAttribute("type");
 
-		if (element != null && !"".equals(element.getTextContent())) {
-			Procedure procedure = new Procedure(element.getTextContent());
-			procedure.getDependencies(master.getWorkspace()).forEach(master::addDependency);
+		if (procedureField != null && procedureField.getTextContent() != null && !"".equals(
+				procedureField.getTextContent())) {
+			Procedure procedure = new Procedure(procedureField.getTextContent());
+			List<Dependency> dependencies = procedure.getDependencies(master.getWorkspace());
 
+			// If the procedure doesn't actually exist, add a warning and skip this block
 			if (!procedure.exists) {
 				master.addCompileNote(new BlocklyCompileNote(BlocklyCompileNote.Type.WARNING,
 						L10N.t("blockly.warnings.call_procedure.nonexistent", procedure.getName())));
 				return;
 			}
 
-			Element x = null, y = null, z = null;
-			List<Element> values = XMLUtil.getChildrenWithName(block, "value");
-			for (Element e : values)
-				switch (e.getAttribute("name")) {
-				case "x":
-					x = e;
-					break;
-				case "y":
-					y = e;
-					break;
-				case "z":
-					z = e;
-					break;
+			// The procedure dependencies, in a flattened {"name": "type"} map
+			Map<String, String> flattenedDeps = dependencies.stream()

```

Comment:
This seems like a lot of duplicated code and logic between ProcedureCallBlock and ProcedureRetvalBlock, something should be done about this, maybe those two could be merged?

Reply:
Common class with two inner subclasses for current classes?

Reply:
Or keep the logic in this class and reuse it in the other?

Reply:
One block can support multiple types, just like procedurecall supports 3 different blocks at the moment. I wouldn't subclass this

Reply:
I think I got what you're asking for :+1:

Reply:
Technically call procedure or call procedure with retval are more or less same thing, one just provides result while one doesn't

Reply:
The main issue is different `BlockType`. How can we get around that one?

Reply:
Fair point. I didn't design it this way. In this case, go with your approach, so one base abstract class "AbstractProcedureCallBlock" or something like this, and children implement as little as needed, including getSupportedBlocks and getBlockType.

It may make sense to do this refactor in a separate PR

Reply:
To this one?

Reply:
To master. I think we could rework this beforehand, don't you think?

Reply:
If there is not enough similarities in master, then making this part of this PR is fine too

Reply:
Hmm, I _would_ say the diff between these files here and on master branch is kind of too large (more noticeable here as I also removed the coordinates part).

Reply:
Ok, then make it part of this PR